cmd/snap-bootstrap/initramfs_install_handler.go

```diff
@@ -0,0 +1,418 @@
+ */
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/asserts/sysdb"
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/gadget"
+	"github.com/snapcore/snapd/gadget/install"
+	"github.com/snapcore/snapd/interfaces/builtin"
+	"github.com/snapcore/snapd/kernel/fde"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+
+	// to set sysconfig.ApplyFilesystemOnlyDefaultsImpl
+	_ "github.com/snapcore/snapd/overlord/configstate/configcore"
+	"github.com/snapcore/snapd/overlord/devicestate"
+	"github.com/snapcore/snapd/progress"
+	"github.com/snapcore/snapd/secboot"
+	"github.com/snapcore/snapd/seed"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/snapfile"
+	"github.com/snapcore/snapd/systemd"
+	"github.com/snapcore/snapd/timings"
+)
+
+const (
+	rootdir          string = "/"
+	InitRamfsFdeHook string = "fde-reveal-key"
+)
+
+var (
+	// assertion store to be trusted
+	trusted = sysdb.Trusted()
+	// /run/mnt/ubuntu-data is not exported, why?
+	InstallUbuntuDataDir = filepath.Join(boot.InitramfsRunMntDir, "ubuntu-data")
+)
+
+// mounts info
+type mountInfo struct {
+	mountPoint string
+	unmountFnc func() error
+	// essential flag is tracking mounts essential for install mode
+	// those mounts should be destroyed when moving to run mode
+	essential bool
+}
+
+// tracking of active mounts
+var activeMounts []mountInfo
+
+func doUnmount(info *mountInfo) error {
+	if err := info.unmountFnc(); err != nil {
+		// only log error and try to unmount next
+		return fmt.Errorf("Failed to unmount%s, %v", info.mountPoint, err)
+	}
+	return nil
+}
+
+func hasFDESetupHook(kernelInfo *snap.Info) (bool, error) {
+	_, err := exec.LookPath(InitRamfsFdeHook)
+	return err == nil, err
+}
+
+func runFDESetupHook(req *fde.SetupRequest) ([]byte, error) {
+	stdin, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf(`cannot build request %v for fde-reveal-key: %v`, req, err)
+	}
+	return fde.RunFDEinitramfsHelper(InitRamfsFdeHook, stdin)
+}
+
+func newMemAssertionsDB(commitObserve func(verified asserts.Assertion)) (db *asserts.Database, commitTo func(*asserts.Batch) error, err error) {
+	memDB, err := asserts.OpenDatabase(&asserts.DatabaseConfig{
+		Backstore: asserts.NewMemoryBackstore(),
+		Trusted:   trusted,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	commitTo = func(b *asserts.Batch) error {
+		return b.CommitToAndObserve(memDB, commitObserve, nil)
+	}
+
+	return memDB, commitTo, nil
+}
+
+func loadDeviceSeed(systemLabel string) (deviceSeed seed.Seed, err error) {
+	deviceSeed, err = seed.Open(boot.InitramfsUbuntuSeedDir, systemLabel)
+	if err != nil {
+		return nil, err
+	}
+	n := runtime.NumCPU()
+	if n > 1 {
+		deviceSeed.SetParallelism(n)
+	}
+	return deviceSeed, err
+}
+
+func loadSeedAssertions(deviceSeed seed.Seed) (*asserts.Database, error) {
+	start := time.Now()
```

Comment:
this gets tedious, we should really be able to reuse timings as it's written to be isolated from state/overlord